Runtime/Behaviours/CinemachineConfiner2D.cs

```diff
@@ -0,0 +1,484 @@
+        /// For more info on frustum height, see <see cref="docs.unity3d.com/Manual/FrustumSizeAtDistance.html"/> 
+        /// </summary>
+        /// <param name="state">CameraState for checking if Orthographic or Perspective</param>
+        /// <param name="vcam">vcam, to check its position</param>
+        /// <returns>Frustum height of the camera</returns>
+        private float CalculateHalfFrustumHeight(in CameraState state, in float cameraPosLocalZ)
+        {
+            float frustumHeight;
+            if (state.Lens.Orthographic)
+            {
+                frustumHeight = state.Lens.OrthographicSize;
+            }
+            else
+            {
+                // distance between the collider's plane and the camera
+                float distance = cameraPosLocalZ;
+                frustumHeight = distance * Mathf.Tan(state.Lens.FieldOfView * 0.5f * Mathf.Deg2Rad);
+            }
+
+            return Mathf.Abs(frustumHeight);
+        }
+        
+        /// <summary>
+        /// Confines input 2D point within the confined area.
+        /// </summary>
+        /// <param name="positionToConfine">2D point to confine</param>
+        /// <returns>Confined position</returns>
+        private Vector2 ConfinePoint(Vector2 positionToConfine, in List<List<Vector2>> pathCache,
+            in bool hasBone, in float windowWidth, in float windowHeight)
+        {
+            if (ShrinkablePolygon.IsInside(pathCache, positionToConfine))
+            {
+                return positionToConfine;
+            }
+            bool outsideOfOriginal = !ShrinkablePolygon.IsInside(m_shapeCache.m_OriginalPath, positionToConfine);
+
+            Vector2 closest = positionToConfine;
+            float minDistance = float.MaxValue;
+            for (int i = 0; i < pathCache.Count; ++i)
+            {
+                int numPoints = pathCache[i].Count;
+                for (int j = 0; j < numPoints; ++j)
+                {
+                    Vector2 v0 = pathCache[i][j];
+                    Vector2 v = pathCache[i][(j + 1) % numPoints];
+                    Vector2 c = Vector2.Lerp(v0, v, positionToConfine.ClosestPointOnSegment(v0, v));
+                    Vector2 difference = positionToConfine - c;
+                    float distance = Vector2.SqrMagnitude(difference);
+                    if (Mathf.Abs(difference.x) > windowWidth || Mathf.Abs(difference.y) > windowHeight)
+                    {
+                        // penalty for points from which the target is not visible, prefering visibility over proximity
+                        distance += m_confinerBaker.m_sqrPolygonDiagonal; 
+                    }
+#if CINEMACHINE_EXPERIMENTAL_CONFINER2D
+                    bool isValid = outsideOfOriginal || !hasBone || !DoesIntersectOriginal(positionToConfine, c);
+#else
+                    bool isValid = outsideOfOriginal || !DoesIntersectOriginal(positionToConfine, c);
```

Comment:
don't you mean !outsideOfOriginal ?

Reply:
No. If it is outside, then it is valid, because I am looking for closest point in that case, I don't care about DoesIntersectOriginal, because if we are outside, we always intersect original.

So two cases:
If outside - ok
If inside, and does not intersect original

Reply:
if Outside: valid
if Inside, and no bone: valid
if Inside, and has bone and does not intersect original: valid